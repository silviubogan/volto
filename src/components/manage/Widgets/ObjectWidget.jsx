import React from 'react';
import { Tab } from 'semantic-ui-react';
import Field from '../Form/Field';
import PropTypes from 'prop-types';

const ObjectWidget = ({
  schema,
  value, // not checked to not contain unknown fields
  onChange,
  errors = {},
  id,
  ...props
}) => {
  const renderFieldSet = React.useCallback(
    (fieldset, fieldsetIndex) => {
      return fieldset.fields.map((field, index) => {
        const myIndex = index;
        console.log('TODO', field, myIndex);
        return (
          <Field
            {...schema.properties[field]}
            id={`${field}-${fieldsetIndex}-${myIndex}`} // if we add here -${id} the form does not work
            fieldset={fieldset.title.toLowerCase()}
            value={value?.[field]}
            required={schema.required.indexOf(field) !== -1}
            onChange={(field, fieldvalue) => {
              return onChange(id, { ...value, [field]: fieldvalue });
            }}
            key={field}
            error={errors?.[field]}
            title={schema.properties[field].title}
          />
        );
      });
    },
    [errors, id, onChange, schema.properties, schema.required, value],
  );

  const createTabPane = React.useCallback(
    (fs, index) => {
      console.log('outer fieldset #', index);
      return (() => {
        const idx = index;
        return () => {
          console.log('inner fieldset #', idx);
          return <Tab.Pane>{renderFieldSet(fs, idx)}</Tab.Pane>;
        };
      })();
    },
    [renderFieldSet],
  );

  const createTab = React.useCallback(
    (fieldset, index) => {
      return {
        menuItem: fieldset.title,
        render: createTabPane(fieldset, index),
      };
    },
    [createTabPane],
  );

  return schema.fieldsets.length === 1 ? (
    renderFieldSet(schema.fieldsets[0], 0)
  ) : (
    <Tab panes={schema.fieldsets.map(createTab)} /> // lazy loading
  );
};

/**
 * Property types.
 * @property {Object} propTypes Property types.
 * @static
 */
ObjectWidget.propTypes = {
  id: PropTypes.string.isRequired,
  schema: PropTypes.object.isRequired,
  errors: PropTypes.object,
  value: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
};

/**
 * Default properties.
 * @property {Object} defaultProps Default properties.
 * @static
 */
ObjectWidget.defaultProps = {};

export default ObjectWidget;
