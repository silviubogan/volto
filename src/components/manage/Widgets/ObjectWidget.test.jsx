import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import ObjectWidget from './ObjectWidget';

const mockStore = configureStore();

const LinkSchema = {
  title: 'Link',
  fieldsets: [
    {
      id: 'internal',
      title: 'Internal',
      fields: ['internal_link'],
    },
    {
      id: 'external',
      title: 'External',
      fields: ['external_link'],
    },
    {
      id: 'email',
      title: 'Email',
      fields: ['email_address', 'email_subject'],
    },
  ],
  properties: {
    internal_link: {
      widget: 'object_browser',
      title: 'Internal link',
    },
    external_link: {
      title:
        'External URL (can be relative within this site or absolute if it starts with http:// or https://)',
    },
    email_address: {
      title: 'Email address',
    },
    email_subject: {
      title: 'Email subject (optional)',
    },
  },
  required: [],
};

test('renders an object widget component', () => {
  const store = mockStore({
    search: {},
    intl: {
      locale: 'en',
      messages: {},
    },
  });
  const component = renderer.create(
    <Provider store={store}>
      <ObjectWidget
        id="my-widget"
        schema={LinkSchema}
        value={{}}
        title="My Widget"
        onChange={() => {}}
        errors={{}}
      />
    </Provider>,
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});

test('renders an object widget controlled component', () => {
  const store = mockStore({
    search: {},
    intl: {
      locale: 'en',
      messages: {},
    },
  });

  const component = renderer.create(
    <Provider store={store}>
      <ObjectWidget
        id="my-widget"
        schema={LinkSchema}
        value={{}}
        title="My Widget"
        onChange={() => {}}
        errors={{}}
      />
    </Provider>,
  );

  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
