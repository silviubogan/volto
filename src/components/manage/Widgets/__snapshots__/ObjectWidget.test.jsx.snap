// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an object widget component 1`] = `
<div>
  <div
    className="ui attached tabular menu"
  >
    <a
      className="active item"
      onClick={[Function]}
    >
      Internal
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      External
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      Email
    </a>
  </div>
  <div
    className="ui bottom attached segment active tab"
  >
    <div
      className="inline field text"
      id="field-internal_link"
    >
      <div
        className="ui grid"
      >
        <div
          className="stretched row"
        >
          <div
            className="four wide column"
          >
            <div
              className="wrapper"
            >
              <label
                htmlFor="field-internal_link"
              >
                Internal link
              </label>
            </div>
          </div>
          <div
            className="eight wide column"
          >
            <div
              className="ui input"
            >
              <input
                disabled={false}
                id="field-internal_link"
                name="internal_link"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders an object widget component with invalid field in value 1`] = `
<div>
  <div
    className="ui attached tabular menu"
  >
    <a
      className="active item"
      onClick={[Function]}
    >
      Internal
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      External
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      Email
    </a>
  </div>
  <div
    className="ui bottom attached segment active tab"
  >
    <div
      className="inline field text"
      id="field-internal_link"
    >
      <div
        className="ui grid"
      >
        <div
          className="stretched row"
        >
          <div
            className="four wide column"
          >
            <div
              className="wrapper"
            >
              <label
                htmlFor="field-internal_link"
              >
                Internal link
              </label>
            </div>
          </div>
          <div
            className="eight wide column"
          >
            <div
              className="ui input"
            >
              <input
                disabled={false}
                id="field-internal_link"
                name="internal_link"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
